assembly/src/parsers/io_ops.rs

```diff
@@ -256,69 +276,137 @@ pub fn parse_loc_storew(op: &Token) -> Result<Node, ParsingError> {
+        .step_by(HEX_CHUNK_SIZE)
+        .map(|i| parse_hex_value(op, &param_str[i..i + HEX_CHUNK_SIZE], 1))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    build_push_many_instruction(values.iter())
+}
+
+/// Parses a hexadecimal parameter value into a u64.
+///
+/// # Errors
+/// Returns an error if:
+/// - The length of hex string is not even.
+/// - The length of hex string is greater than 16.
+/// - If the string does not contain a valid hexadecimal value.
+/// - If the parsed value is greater than or equal to the field modulus.
+fn parse_hex_value(op: &Token, param_str: &str, param_idx: usize) -> Result<u64, ParsingError> {
+    if param_str.len() % 2 != 0 {
+        return Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            &format!("hex string '{param_str}' does not contain an even number of characters"),
+        ));
+    }
+
+    if param_str.len() > HEX_CHUNK_SIZE {
+        return Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            &format!("hex string '{param_str}' contains too many characters"),
+        ));
+    }
+
+    let value = u64::from_str_radix(param_str, 16)
+        .map_err(|_| ParsingError::invalid_param(op, param_idx))?;
+
+    if value >= Felt::MODULUS {
+        Err(ParsingError::invalid_param_with_reason(
+            op,
+            param_idx,
+            &format!("hex string '{param_str}' contains value greater than field modulus"),
+        ))
     } else {
-        // parse 1 decimal value and push it onto the stack
-        let value = parse_decimal_param(op, param_idx, param_str)?;
-        constants.push(value);
+        Ok(value)
     }
+}
 
-    Ok(constants)
+/// Determines the minimal type appropriate for provided value and returns appropriate instruction
+/// for this value
+fn build_push_one_instruction(value: u64) -> Result<Node, ParsingError> {
+    if value <= u8::MAX as u64 {
+        Ok(Instruction(PushU8(value as u8)))
+    } else if value <= u16::MAX as u64 {
+        Ok(Instruction(PushU16(value as u16)))
+    } else if value <= u32::MAX as u64 {
+        Ok(Instruction(PushU32(value as u32)))
+    } else if value < Felt::MODULUS {
+        Ok(Instruction(PushFelt(Felt::new(value))))
+    } else {
+        unreachable!()
+    }
 }
 
-fn parse_hex_params(
-    op: &Token,
-    param_idx: usize,
-    param_str: &str,
-) -> Result<Vec<Felt>, ParsingError> {
-    // handle error cases where the hex string is poorly formed
-    let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
-        if param_str.len() % 2 != 0 {
-            // parameter string is not a valid hex representation
-            return Err(ParsingError::invalid_param(op, param_idx));
+/// Determines the minimal type appropriate for provided values iterator and returns appropriate
+/// instruction for this values
+fn build_push_many_instruction(values: core::slice::Iter<u64>) -> Result<Node, ParsingError> {
```

Comment:
We always take Iterator as generic so any iterator can fit.

The function should be `fn build_push_many_instruction<I>(values: I) -> Result<Node, ParsingError> where I: Iterator<Item = u64>`